README
NAME
    RTx::EmailCompletion - Auto completion for email field

VERSION
    This document describes version 0.01 of RTx::EmailCompletion

DESCRIPTION
    I'm so tired to type email address by hand that I've done this module
    try Add ajax autocompletion on email field

    There's 3 things :

    *   a small web service "html/Ajax/Email" which search in all known
        users in the database

    *   the scriptaculous library to autocomplete

    *   and modified html RT pages to use those two first component

    As I've embedded some pages of RT, it should work only for release
    3.6.3. But I keep the patches of all modified files in "patches/" so you
    can try to use them against other release.

INSTALLATION
    Install it like a standard perl module

     perl Makefile.PL
     make
     make install

AUTHORS
    Nicolas Chuche <nchuche@barna.be>

COPYRIGHT
    Copyright 2007 by Nicolas Chuche <nchuche@barna.be>

    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.

    See <http://www.perl.com/perl/misc/Artistic.html>

    All Scriptaculous and Prototype program are placed under MIT licence and
    are copyrighted by their owners (see top of files).



CHANGES
Changes for 0.05

  * bugfix for RT < 3.8 : add prototype.js

Changes for 0.04

  * compatibility with RT 3.8
  * deploy scriptaculous only for RT < 3.8
  * use Module::Install::RTx 0.23
  * add newest scriptaculous

Changes for 0.03

  * allow searching on fields other than email (Kevin Falcone && Best Practical)
  * LDAP autocompletion (emails LDAP background color in light grey)
  * optimisation in the match privileged/unprivileged criteria (faster than the if ... elsif ... else inside the loop)

Changes for 0.02

  * allow email completion for unprivileged users (disable by default). See README
  * add debugging tools to find input tags found by emailcompletion.js

Changes for 0.01

  * first release

inc/Module/Install/RTx.pm
#line 1 "inc/Module/Install/RTx.pm - /usr/local/share/perl/5.8.3/Module/Install/RTx.pm"
# $File: //member/autrijus/Module-Install-RTx/lib/Module/Install/RTx.pm $ $Author: autrijus $
# $Revision: #15 $ $Change: 10598 $ $DateTime: 2004/05/13 01:27:54 $ vim: expandtab shiftwidth=4

package Module::Install::RTx;
use Module::Install::Base; @ISA = qw(Module::Install::Base);

$Module::Install::RTx::VERSION = '0.07';

use strict;
use FindBin;
use File::Glob;
use File::Basename;

sub RTx {
    my ($self, $name) = @_;
    my $RTx = 'RTx';
    $RTx = $1 if $name =~ s/^(\w+)-//;

    $self->name("$RTx-$name")
        unless $self->name;
    $self->abstract("RT $name Extension")
        unless $self->abstract;
    $self->version_from (-e "$name.pm" ? "$name.pm" : "lib/$RTx/$name.pm")
        unless $self->version;

    my @prefixes = (qw(/opt /usr/local /home /usr /sw ));
    my $prefix = $ENV{PREFIX};
    @ARGV = grep { /PREFIX=(.*)/ ? (($prefix = $1), 0) : 1 } @ARGV;

    if ($prefix) {
        $RT::LocalPath = $prefix;
        $INC{'RT.pm'} = "$RT::LocalPath/lib/RT.pm";
    }
    else {
        local @INC = (
            @INC,
            $ENV{RTHOME},
            map {( "$_/rt3/lib", "$_/lib/rt3", "$_/lib" )} grep $_, @prefixes
        );
        until ( eval { require RT; $RT::LocalPath } ) {
            warn "Cannot find the location of RT.pm that defines \$RT::LocalPath. ($@)\n";
            $_ = $self->prompt("Path to your RT.pm:") or exit;
            push @INC, $_, "$_/rt3/lib", "$_/lib/rt3";
        }

        package RT;
        RT::LoadConfig();
    }

    my $lib_path = dirname($INC{'RT.pm'});
    print "Using RT configurations from $INC{'RT.pm'}:\n";

    $RT::LocalVarPath	||= $RT::VarPath;
    $RT::LocalPoPath	||= $RT::LocalLexiconPath;
    $RT::LocalHtmlPath	||= $RT::MasonLocalComponentRoot;

    my %path;
    my $with_subdirs = $ENV{WITH_SUBDIRS};
    @ARGV = grep { /WITH_SUBDIRS=(.*)/ ? (($with_subdirs = $1), 0) : 1 } @ARGV;
    my %subdirs = map { $_ => 1 } split(/\s*,\s*/, $with_subdirs);

    foreach (qw(bin etc html po sbin var)) {
        next unless -d "$FindBin::Bin/$_";
        next if %subdirs and !$subdirs{$_};
        $self->no_index( directory => $_ );

        no strict 'refs';
        my $varname = "RT::Local" . ucfirst($_) . "Path";
        $path{$_} = ${$varname} || "$RT::LocalPath/$_";
    }

    $path{$_} .= "/$name" for grep $path{$_}, qw(etc po var);
    my $args = join(', ', map "q($_)", %path);
    $path{lib} = "$RT::LocalPath/lib" unless %subdirs and !$subdirs{'lib'};
    print "./$_\t=> $path{$_}\n" for sort keys %path;

    if (my @dirs = map { (-D => $_) } grep $path{$_}, qw(bin html sbin)) {
        my @po = map { (-o => $_) } grep -f, File::Glob::bsd_glob("po/*.po");
        $self->postamble(<< ".") if @po;
lexicons ::
\t\$(NOECHO) \$(PERL) -MLocale::Maketext::Extract::Run=xgettext -e \"xgettext(qw(@dirs @po))\"
.
    }

    my $postamble = << ".";
install ::
\t\$(NOECHO) \$(PERL) -MExtUtils::Install -e \"install({$args})\"
.

    if ($path{var} and -d $RT::MasonDataDir) {
        my ($uid, $gid) = (stat($RT::MasonDataDir))[4, 5];
        $postamble .= << ".";
\t\$(NOECHO) chown -R $uid:$gid $path{var}
.
    }

    my %has_etc;
    if (File::Glob::bsd_glob("$FindBin::Bin/etc/schema.*")) {
        # got schema, load factory module
        $has_etc{schema}++;
        $self->load('RTxFactory');
        $self->postamble(<< ".");
factory ::
\t\$(NOECHO) \$(PERL) -Ilib -I"$lib_path" -Minc::Module::Install -e"RTxFactory(qw($RTx $name))"

dropdb ::
\t\$(NOECHO) \$(PERL) -Ilib -I"$lib_path" -Minc::Module::Install -e"RTxFactory(qw($RTx $name drop))"

.
    }
    if (File::Glob::bsd_glob("$FindBin::Bin/etc/acl.*")) {
        $has_etc{acl}++;
    }
    if (-e 'etc/initialdata') {
        $has_etc{initialdata}++;
    }

    $self->postamble("$postamble\n");
    if (%subdirs and !$subdirs{'lib'}) {
        $self->makemaker_args(
            PM => { "" => "" },
        )
    }
    else {
        $self->makemaker_args( INSTALLSITELIB => "$RT::LocalPath/lib" );
    }

    if (%has_etc) {
        RTxInit();
        $RT::SbinPath = $RT::LocalPath;
        $RT::SbinPath =~ s/local$/sbin/;

        print "For first-time installation, type 'make initdb'.\n";
        my $initdb = "initdb ::\n";
        $initdb .= <<"." if $has_etc{schema};
\t\$(NOECHO) \$(PERL) -Ilib -I"$lib_path" "$RT::SbinPath/rt-setup-database" --action=schema --datadir ./etc/ --prompt-for-dba-password --dba $RT::DatabaseUser
.
        $initdb .= <<"." if $has_etc{acl};
\t\$(NOECHO) \$(PERL) -Ilib -I"$lib_path" "$RT::SbinPath/rt-setup-database" --action=acl --datadir ./etc/ --prompt-for-dba-password --dba $RT::DatabaseUser
.
        $initdb .= <<"." if $has_etc{initialdata};
\t\$(NOECHO) \$(PERL) -Ilib -I"$lib_path" "$RT::SbinPath/rt-setup-database" --action=insert --datafile=etc/initialdata --dba $RT::DatabaseUser
.
        $self->postamble("$initdb\n");
    }
}

sub RTxInit {
    unshift @INC, substr(delete($INC{'RT.pm'}), 0, -5) if $INC{'RT.pm'};
    require RT;
    RT::LoadConfig();
    RT::ConnectToDatabase();

    die "Cannot load RT" unless $RT::Handle and $RT::DatabaseType;
}

1;

__END__

#line 238


extra/bench_search_rdbms.pl
#!/usr/bin/perl -w

use strict;
use lib qw(/opt/rt3/lib/);
use Benchmark qw/:all/;
use Data::Dumper;

use RT;
RT::LoadConfig();
RT::Init();



my $CurrentUser =  RT::User->new($RT::SystemUser);

$CurrentUser->LoadByEmail('Nicolas.Chuche@equipement.gouv.fr');
# $CurrentUser->LoadByEmail('Eric.Boyon@equipement.gouv.fr');

my $Users = RT::Users->new( $CurrentUser );

$RT::EmailCompletionNonPrivileged ||= '';
$RT::EmailCompletionSearchFields  ||= [qw(EmailAddress)];
$RT::EmailCompletionSearch        ||= 'LIKE';

#$RT::EmailCompletionUnprivileged = "privileged";

my $param = "ch";

print "Test pour NBC\n";
{
    $CurrentUser->LoadByEmail('Nicolas.Chuche@equipement.gouv.fr');
    timethese(10, {'Old' => sub { search_rdbms($param, $CurrentUser);  }, 'New' => sub { search_rdbms_new($param, $CurrentUser);  }, });
}

print "Test pour ERIC\n";
{
    $CurrentUser->LoadByEmail('Eric.Boyon@equipement.gouv.fr');
    timethese(10, {'Old' => sub { search_rdbms($param, $CurrentUser);  }, 'New' => sub { search_rdbms_new($param, $CurrentUser);  }, });
}

$RT::EmailCompletionUnprivileged = "privileged";

print "Test pour NBC privileged\n";
{
    $CurrentUser->LoadByEmail('Nicolas.Chuche@equipement.gouv.fr');
    timethese(10, {'Old' => sub { search_rdbms($param, $CurrentUser);  }, 'New' => sub { search_rdbms_new($param, $CurrentUser);  }, });
}

print "Test pour ERIC privileged\n";
{
    $CurrentUser->LoadByEmail('Eric.Boyon@equipement.gouv.fr');
    timethese(10, {'Old' => sub { search_rdbms($param, $CurrentUser);  }, 'New' => sub { search_rdbms_new($param, $CurrentUser);  }, });
}

$RT::EmailCompletionUnprivileged = qr/gouv.fr/i;

print "Test pour NBC regexp\n";
{
    $CurrentUser->LoadByEmail('Nicolas.Chuche@equipement.gouv.fr');
    timethese(10, {'Old' => sub { search_rdbms($param, $CurrentUser);  }, 'New' => sub { search_rdbms_new($param, $CurrentUser);  }, });
}

print "Test pour ERIC regexp\n";
{
    $CurrentUser->LoadByEmail('Eric.Boyon@equipement.gouv.fr');
    timethese(10, {'Old' => sub { search_rdbms($param, $CurrentUser);  }, 'New' => sub { search_rdbms_new($param, $CurrentUser);  }, });
}


sub search_rdbms {
    my $Email = shift;
    my $CurrentUser = shift;

    my @emails;

    my $Users = RT::Users->new( $RT::SystemUser );
    foreach my $field (@{$RT::EmailCompletionSearchFields}) {
	$Users->Limit(SUBCLAUSE => 'EmailCompletion', ALIAS => 'main', FIELD => $field, OPERATOR => $RT::EmailCompletionSearch, VALUE => $Email, ENTRYAGGREGATOR => 'OR');
    }

    $RT::Logger->debug($Users->BuildSelectQuery);

    while (my $User = $Users->Next()) {
	next if $User->id == $RT::Nobody->id;

	# some cleaning on emailaddress
	next if $User->EmailAddress !~ m{[a-zA-Z0-9_.-]+@[^.]+\.[^.]};
	next if $User->EmailAddress =~ m{[,?!/;\\]};

	# if you're privileged user you can see anybody
	#
	# if you're not by default you can see nobody
	# if RT::EmailCompletionUnprivileged is set to anybody you can see anybody
	# else you can see only privileged users

	if ( $CurrentUser->Privileged() or $RT::EmailCompletionUnprivileged eq 'everybody' ) {
	    #print "privileged\n";
	    # Ok, show everybody
	} elsif ( $RT::EmailCompletionUnprivileged eq 'privileged' ) {
	    #print "unpriv and only priv users\n";
	    next unless $User->Privileged();
	} elsif ( ref($RT::EmailCompletionUnprivileged) eq 'Regexp' ) {
	    #print "regexp\n";
	    next unless $User->EmailAddress =~ m/$RT::EmailCompletionUnprivileged/;
	} else {
	    #print "other\n";
	    next
	}

	push @emails, $User->EmailAddress;
    }
    @emails;
}

sub search_rdbms_new {
    my $Email = shift;
    my $CurrentUser = shift;

    return unless $CurrentUser->Privileged() or defined $RT::EmailCompletionUnprivileged;

    my $Users = RT::Users->new( $RT::SystemUser );
    foreach my $field (@{$RT::EmailCompletionSearchFields}) {
	$Users->Limit(SUBCLAUSE => 'EmailCompletion', ALIAS => 'main', FIELD => $field, OPERATOR => $RT::EmailCompletionSearch, VALUE => $Email, ENTRYAGGREGATOR => 'OR');
    }

    $RT::Logger->debug($Users->BuildSelectQuery);

    my @users;
    while (my $User = $Users->Next()) {
	next if $User->id == $RT::Nobody->id;

	# some cleaning on emailaddress
	next if $User->EmailAddress !~ m{[a-zA-Z0-9_.-]+@[^.]+\.[^.]};
	next if $User->EmailAddress =~ m{[,?!/;\\]};

	push @users, $User;
    }

    # if you're privileged user you can see anybody
    #
    # if you're not by default you can see nobody
    # if RT::EmailCompletionUnprivileged is set to anybody you can see anybody
    # else you can see only privileged users

    if ( $CurrentUser->Privileged() or $RT::EmailCompletionUnprivileged eq 'everybody' ) {
	# Ok, show everybody

    } elsif ( $RT::EmailCompletionUnprivileged eq 'privileged' ) {
	@users = grep { $_->Privileged()  } @users;

    } elsif ( ref($RT::EmailCompletionUnprivileged) eq 'Regexp' ) {
	@users = grep { $_->EmailAddress() =~ m/$RT::EmailCompletionUnprivileged/ } @users;

    } else {
	@users = ();
    }

    my @email = map { $_->EmailAddress() } @users;
}


lib/RTx/EmailCompletion/Ldap.pm
package RTx::EmailCompletion::Ldap;

use Net::LDAP;
use Net::LDAP::Constant qw(LDAP_SUCCESS LDAP_PARTIAL_RESULTS);
use Net::LDAP::Util qw (ldap_error_name);
use RT::Users;

sub search_ldap {
    my $param = shift;
    my $CurrentUser = shift;

    return if length($param) < $RT::EmailCompletionLdapMinLength;

    # if user isn't privileged and we want only show privileged user, return now
    return if not $CurrentUser->Privileged() and $RT::EmailCompletionUnprivileged eq 'privileged';

    my $ldap = new Net::LDAP($RT::EmailCompletionLdapServer);

    my $mesg = defined $RT::EmailCompletionLdapUser && $RT::EmailCompletionLdapUser ne '' ?
	$ldap->bind($RT::EmailCompletionLdapUser, password => $RT::EmailCompletionLdapPass)
	    : $ldap->bind();

    if ($mesg->code != LDAP_SUCCESS) {
	$RT::Logger->crit("Unable to bind to $RT::EmailCompletionLdapServer: ", ldap_error_name($mesg->code), "\n");
	return;
    }

    my $filter = "(|" . join('', map { "($_=*$param*)" } @{ $RT::EmailCompletionLdapAttrSearch }) . ")";
    $filter = "(&" . $RT::EmailCompletionLdapFilter . $filter . ")" if $RT::EmailCompletionLdapFilter;

    $RT::Logger->debug("LDAP filter is: $filter\n") if RTx::EmailCompletion::DEBUG;

    $mesg = $ldap->search(base   => $RT::EmailCompletionLdapBase,
			  filter => $filter,
			  attrs  => $RT::EmailCompletionLdapAttrShow);

    if ($mesg->code != LDAP_SUCCESS and $mesg->code != LDAP_PARTIAL_RESULTS)  {
	$RT::Logger->crit("Unable to search in LDAP: ", ldap_error_name($mesg->code), "\n");
    }

    my @emails = map { $_->get_attribute( $RT::EmailCompletionLdapAttrShow ) } $mesg->entries;

    @emails = grep { m/$RT::EmailCompletionUnprivileged/ } @emails
	if ref($RT::EmailCompletionUnprivileged) eq 'Regexp' and not $CurrentUser->Privileged();

    $mesg = $ldap->unbind();
    if ($mesg->code != LDAP_SUCCESS) {
	$RT::Logger->crit("Unable to unbind from LDAP:", ldap_error_name($mesg->code), "\n");
	return;
    }

    $RT::Logger->debug("emails returned are: @emails\n") if RTx::EmailCompletion::DEBUG;

    sort @emails;
}

1;


html/NoAuth/js/emailcompletion.js
function initPage() {
    addAutoComplete();
} // initPage

/* CUSTOMIZE PART */

// documentation can be found in lib/RTx/EmailCompletion.pm

// we need two Arrays :
// one for input with multiple email address allowed
var multipleCompletion = new Array("Requestors", "Cc", "AdminCc", "WatcherAddressEmail[123]", "UpdateCc", "UpdateBcc");

// and one for input with only one email address allowed
var singleCompletion   = new Array("(Add|Delete)Requestor", "(Add|Delete)Cc", "(Add|Delete)AdminCc");

/* END OF CUSTOMIZE PART */

var globalRegexp   = new RegExp('^(' + multipleCompletion.concat(singleCompletion).join('|') + ')$');
var multipleRegexp = new RegExp('^(' + multipleCompletion.join('|') + ')$');

function addAutoComplete() {
    var inputs = document.getElementsByTagName("input");

    for (var i = 0; i < inputs.length; i++) {
	var input = inputs[i];
	var inputName = input.getAttribute("name");

	// if empty or not defined
	if (! inputName)
	    continue;

	// only input's names defined in global vars at the beginning
	// are concerned
	if ( ! inputName.match(globalRegexp) )
	    continue;

	// if multiple email address are allowed we add the token
	// option to Autocompleter
	var options = '';
	if (inputName.match(multipleRegexp))
	    options = "tokens: ','";

	// we must set an id for scriptaculous
	// we use the name that seems to be uniq
	input.id = inputName;

	// DEBUGGING PURPOSE
	// input.className += input.className? " emailcompletiondebug" : "emailcompletiondebug";

	var div = '<div class="autocomplete" id="' + inputName + '_to_auto_complete" />';
	div += '<script type="text/javascript">new Ajax.Autocompleter(\'' + inputName;
	div += "', '" + inputName + "_to_auto_complete', '<%$RT::WebPath%>/SelfService/Ajax/EmailCompletion\', {" + options + "})</script>";

	// use prototype to add the div after input
	new Insertion.After(inputName,div);
    }

} //addAutoComplete

Event.observe(window, 'load', initPage);


lib/RTx/EmailCompletion.pm
package RTx::EmailCompletion;

use strict;
use RT::Users;

our $VERSION = "0.05";

use constant DEBUG => 0;

sub search_rdbms {
    my $Email = shift;
    my $CurrentUser = shift;

    return unless $CurrentUser->Privileged() or defined $RT::EmailCompletionUnprivileged;

    my @emails;

    my $Users = RT::Users->new( $CurrentUser );
    foreach my $field (@{$RT::EmailCompletionSearchFields}) {
	$Users->Limit(SUBCLAUSE => 'EmailCompletion', ALIAS => 'main', FIELD => $field, OPERATOR => $RT::EmailCompletionSearch, VALUE => $Email, ENTRYAGGREGATOR => 'OR');
    }

    $RT::Logger->debug($Users->BuildSelectQuery);

    my @users;
    while (my $User = $Users->Next()) {
	next if $User->id == $RT::Nobody->id;

	# some cleaning on emailaddress
	next if $User->EmailAddress !~ m{[a-zA-Z0-9_.-]+@[^.]+\.[^.]};
	next if $User->EmailAddress =~ m{[,?!/;\\]};

	push @users, $User;
    }

    # if you're privileged user you can see anybody
    #
    # if you're not by default you can see nobody
    # if RT::EmailCompletionUnprivileged is set to anybody you can see anybody
    # else you can see only privileged users

    if ( $CurrentUser->Privileged() or $RT::EmailCompletionUnprivileged eq 'everybody' ) {
	# Ok, show everybody

    } elsif ( $RT::EmailCompletionUnprivileged eq 'privileged' ) {
	@users = grep { $_->Privileged()  } @users;

    } elsif ( ref($RT::EmailCompletionUnprivileged) eq 'Regexp' ) {
	@users = grep { $_->EmailAddress() =~ m/$RT::EmailCompletionUnprivileged/ } @users;

    } else {
	@users = ();
    }

    my @email = map { $_->EmailAddress() } @users;
}

# we dynamically build search function

our $AUTOLOAD;
sub AUTOLOAD {
    (my $function = $AUTOLOAD) =~ s/.*:://;
    die "Unable to find search function in AUTOLOAD" unless $function eq 'search';

    my $mod_ldap;
    if ($RT::EmailCompletionLdapServer and not $RT::EmailCompletionLdapDisabled) {
	eval {
	    require RTx::EmailCompletion::Ldap;
	};
	if ($@) {
	    $RT::Logger->crit("Unable to load RTx::EmailCompletion::Ldap, perhaps you forgot to install Net::LDAP: $@\n");
	} else {
	    $mod_ldap = 1;
	}
    }
    my $str = 'sub search { my (@emails, @ldaps);';
    $str   .= '@emails = search_rdbms(@_);'                            unless $RT::EmailCompletionRdbmsDisabled;
    $str   .= '@ldaps  = RTx::EmailCompletion::Ldap::search_ldap(@_);' if $mod_ldap;
    $str   .= 'return (\@emails, \@ldaps); }';

    $RT::Logger->debug("function used is $str\n") if DEBUG;

    eval $str;
    goto &search;
}


1;

__END__

=head1 NAME

RTx::EmailCompletion - Add auto completion on RT email fields

=head1 VERSION

This document describes version 0.05 of RTx::EmailCompletion.

=head1 DESCRIPTION

I'm so tired to type email address by hand that I've done this module
to add AJAX autocompletion on all email field of RT. As adding
completion is dynamic, it should work on most RT releases (see later
if it's not the case).

There's 4 S<things :>

=over

=item *

a small web service C<html/Ajax/EmailCompletion> which search in all
known users in the S<database ;>

=item *

prototype library to manipulate DOM and scriptaculous library to
S<autocomplete ;>

=item *

a small javascript which parse html pages and add autocomplete on
known input tags.

=item *

a perl module to handle all the logic

=back

=head1 INSTALLATION

if upgrading from 0.01, see later UPGRADE FROM 0.01.

if you've already use an earlier version of this module with RT >=
3.8, see later SCRIPTACULOUS AND RT 3.8

Install it like a standard perl module :

 RTHOME=/opt/rt3 perl Makefile.PL
 make
 make install

=head1 CONFIGURATION

This section is fairly long but you don't really need to read it if
you just want the basic : autocompletion only for privileged users
against all registred users of RT database.

=head2 unprivileged users autocompletion

By default, completion works only for privileged users.

You can activate it for unprivileged users (in the SelfService) by
setting $EmailCompletionUnprivileged in
F<RTHOME/etc/RT_SiteConfig.pm>. There's three ways :

=over

=item *
show everybody

  Set($EmailCompletionUnprivileged,"everybody");

B<Be careful>, this will also show all yours LDAP users.

=item *
show only privileged users

  Set($EmailCompletionUnprivileged,"privileged");

This won't show LDAP users

=item *
show only email matching a regexp

  Set($EmailCompletionUnprivileged, qr/\@my\.corp\.domain$/ );

This will also show LDAP user mails that matchs the regexp

=back

=head2 change the database clause to search email

You can also change the operator used in the C<where> clause to search
email with the global var $RT::EmailCompletionSearch. The default one
is C<LIKE>.

To change it, add a line like this in F<RTHOME/etc/RT_SiteConfig.pm> :

 Set($EmailCompletionSearch, "STARTSWITH");

This variable can take the values C<LIKE>, C<STARTSWITH> and
C<ENDSWITH>.

By default, the plugin searches on Users.EmailAddress.

You can change where it searches by setting $EmailCompletionSearchFields
in RTHOME/etc/RT_SiteConfig.pm to an arrayref of fields from the
Users table.

   Set( $EmailCompletionSearchFields, [qw(EmailAddress RealName Name)] );

This would allow you to search by usernames, full names and email addresses

=head2 LDAP configuration

Starting with RTx::EmailCompletion 0.03, autocompletion works with LDAP
servers.

If you already have installed and configured LDAP authentication
overlay, this configuration will be used and it should/could work just
as it is.

The following configuration parameters applied :

=over

=item *
EmailCompletionLdapServer : the ldap server (mandatory)

  Set($EmailCompletionLdapServer, "my.ldap.server");

If not set, RTx::EmailCompletion will search for LdapServer parameter
(configured for the LDAP RT authentification layout and some others
LDAP RT extensions).

=item *
EmailCompletionLdapBase : the ldap base (mandatory)

  Set($EmailCompletionLdapBase, "dc=debian,dc=org");

If not set, RTx::EmailCompletion will search for LdapBase parameter
(configured for the LDAP RT authentification layout and some others
LDAP RT extensions).

=item *
EmailCompletionLdapUser : the ldap user if you need authentication

  Set($EmailCompletionLdapUser, "myldapuser");

If not set, RTx::EmailCompletion will search for LdapUser parameter
(configured for the LDAP RT authentification layout and some others
LDAP RT extensions).

=item *
EmailCompletionLdapPass : the ldap password if you need authentication

  Set($EmailCompletionLdapPass, "mypassword");

If not set, RTx::EmailCompletion will search for LdapPass parameter
(configured for the LDAP RT authentification layout and some others
LDAP RT extensions).

=item *
EmailCompletionLdapFilter : the ldap filter if needed

  Set($EmailCompletionLdapFilter, "(objectclass=person)");

If not set, RTx::EmailCompletion will search for LdapFilter parameter
(configured for the LDAP RT authentification layout).

=item *
EmailCompletionLdapAttrSearch : the ldap search attributes

  Set($EmailCompletionLdapAttrSearch, [qw/mail cn/]);

Default value is mail.

=item *
EmailCompletionLdapAttrShow : the mail attribute name

  Set($EmailCompletionLdapAttrShow, "mail");

Default value is mail

=item *
EmailCompletionLdapMinLength : minimum parameter length to send an ldap request

  Set(EmailCompletionLdapMinLength, 6);

Default value is 4

=back

The minimum LDAP configuration look somethink like this :

  Set($EmailCompletionLdapServer, "db.debian.org");
  Set($EmailCompletionLdapBase, "dc=debian,dc=org");

You can disable ldap completion (useful if you have installed ldap
authentication overlay and you don't want ldap completion) with :

  Set($EmailCompletionLdapDisabled, 1);

If you want to keep only LDAP completion, you can also disable RDBMS :

  Set($EmailCompletionRdbmsDisabled, 1);

The given value must be true for perl.

=head1 HOW TO ADD FIELD TO AUTOCOMPLETION

If you find email field without autocomplete, you can modify
C<html/NoAuth/js/emailcompletion.js> to handle this field (and email
me to patch this module).

At the beginning of this file you will find two global vars
C<multipleCompletion> and C<singleCompletion>. They are array of
regexp.

Regexp must match all the word because C<^> and C<$> are added for
matching. So if you want to match C<Field1> and C<Field2> you must add
something like C<Field.> or better C<Field[12]>.

To verify that javascript find your input tag, you can uncomment the
line just after the "DEBUGGING PURPOSE" one. All input tags find by
the script will appear with a big red border.

=head1 UPGRADE FROM 0.01

As directory structure has change, If you upgrade from 0.01, you must
delete :

  RTHOME/local/html/Ajax/EmailCompletion
  RTHOME/local/html/NoAuth/js/emailcompletion.js
  RTHOME/local/html/NoAuth/js/
  RTHOME/local/html/NoAuth/css/emailcompletion.css

B<Be careful> if you have other javascripts in RTHOME/local/html/NoAuth/js/

=head1 SCRIPTACULOUS AND RT 3.8

RT 3.8 and later come with their own scriptaculous library. So if you
have installed an earlier release of this module with RT 3.8 and
later, the original scriptaculous will be overriden by this module.

The best solution is to remove RTHOME/share/html/NoAuth/js/ and copy
RTSOURCE/share/html/NoAuth/js/ to RTHOME/share/html/NoAuth/.

This release is not subject to this problem.

=head1 HISTORY

The first version (unreleased) modify html pages. The better method
actually used allow this module to be compatible with, virtually, all
RT release.

=head1 AUTHORS

Nicolas Chuche E<lt>nchuche@barna.beE<gt>

=head1 COPYRIGHT

Copyright 2007 by Nicolas Chuche E<lt>nchuche@barna.beE<gt>

This program is free software; you can redistribute it and/or
modify it under the same terms as Perl itself.

See L<http://www.perl.com/perl/misc/Artistic.html>

All Scriptaculous and Prototype program are placed under MIT licence
and are copyrighted by their owners (see top of files).

=cut
